cpp/src/gandiva/precompiled/string_ops.cc

```diff
@@ -191,38 +191,57 @@ VAR_LEN_TYPES(IS_NOT_NULL, isnotnull)
 #undef IS_NOT_NULL
 
 FORCE_INLINE
-const char* substr_utf8_int64_int64(int64 context, const char* input, int32 in_len,
-                                    int64 offset64, int64 length, int32* out_len) {
-  if (length <= 0 || input == nullptr || in_len <= 0) {
-    *out_len = 0;
+const char* substr_utf8_int64_int64(int64 context, const char* input, int32 in_data_len,
+                                    int64 offset64, int64 length, int32* out_data_len) {
+  if (length <= 0 || input == nullptr || in_data_len <= 0) {
+    *out_data_len = 0;
     return "";
   }
 
+  int32 num_in_chars;  // num of chars in input string
+  int32* char_start =
+      reinterpret_cast<int32*>(malloc(in_data_len * 4));  // stores start index of chars
+
+  int32 curr_char_pos = 0;
+  char curr_char_num_bytes = 0;
+
+  char_start[curr_char_pos] = 0;
+  for (int32 i = 0; i < in_data_len; i = i + curr_char_num_bytes) {
```

Comment:
There's a potential out of bound issue if the in_data_len = 2^31 -1 and the last utf8 byte is the start of a glyph with more that 1 character. Also, this is an infinite loop (+ out of bound buffer write) if it encounters a single invalid byte.

Disclaimer, I haven't tested this.

```C++

int32_t start_pos, end_pos = -1;

int32_t n_glyphs = 0;
int64_t pos = 0;
while (pos < static_cast<int64_t>(in_data_len)) {
  if (n_glyphs == offset) start_pos = pos;
  if (n_glyphs - offset + 1 == length) end_pos = pos; 

  n_glyphs++;
  int64_t glyph_len = utf8_char_length(input[pos])
  // Skip invalid glyphs.
  if (glyph_len == 0) {
    // should it decrease n_glyphs?
    glyph_len = 1;
  }
  pos += glyph_len;
}

// [start_post, end_pos) contains what you want.
```

Reply:
you're right but this will happen if the input string is half a billion characters wide

Reply:
updated. please take a look

Reply:
@fsaintjacques

Reply:
Ok, will give a look.